### DISCUSSION TITLE
Node Shape issue when updating diagram using UpdateModelAction

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#440] [Q&A] [answered]
I am using GLSP vscode extension(https://github.com/eclipse-glsp/glsp-vscode-integration)  + GLSP server.

I am building graph using GModel (org.eclipse.glsp.graph.GGraph > org.eclipse.glsp.graph.impl.GNodeImpl > org.eclipse.glsp.graph.impl.GLabelImpl).

**Screenshot:**
![image](https://user-images.githubusercontent.com/59145404/139389402-e608e412-a0e9-4ee5-a8c5-5ca915887904.png)

When name of node is edited in the text file, we are triggering update model action(org.eclipse.glsp.server.features.core.model.UpdateModelAction.UpdateModelAction(GModelRoot, boolean)). We are generating new graph object(GGraph) and updating root element using UpdateModelAction.

Name will be updated in the diagram but node shape will not be retained .**Ids of node and labels are same for both Graphs.** 
**Screenshot:** Node name changed from new to new1 in the text editor
![image](https://user-images.githubusercontent.com/59145404/139390007-27aa00c1-8c35-4096-bfea-38f5c641270f.png)

**If we perform layout operation on graph (alt + l) above layout issue will be fixed.**

Steps to reproduce the issue:
**We are triggering diagram update from xtext(sync between diagram and text editor). But we can reproduce this issue with below steps without xtext.**
1. In org.eclipse.glsp.server.features.core.model.JsonFileGModelLoader class use below code to generate simple node(loadSourceModel() method) instead of generating graph object from gson parser.

		GGraph graph = new GGraphImpl();
		graph.setId(DefaultTypes.GRAPH + Integer.toString(0));
		graph.setType(DefaultTypes.GRAPH);
		GNodeImpl gNode = new GNodeImpl();
		gNode.setParent(graph);
		gNode.setId("node");
		gNode.setType("node");
		gNode.getCssClasses().add("node");
		gNode.setLayout(GConstants.Layout.VBOX);
		GDimensionImpl d = new GDimensionImpl();
		d.setWidth(50);
		d.setHeight(50);
		gNode.setSize(d);
		
2. org.eclipse.glsp.server.features.modelsourcewatcher.FileWatcher.notifyClient(ClientNotification) here call update action with below code.
		GGraph graph = new GGraphImpl();
		graph.setId(DefaultTypes.GRAPH + Integer.toString(0));
		graph.setType(DefaultTypes.GRAPH);
		GNodeImpl gNode = new GNodeImpl();
		gNode.setParent(graph);
		gNode.setId("node");
		gNode.setType("node");
		gNode.getCssClasses().add("node");
		gNode.setLayout(GConstants.Layout.VBOX);
		GDimensionImpl d = new GDimensionImpl();
		d.setWidth(50);
		d.setHeight(50);
		gNode.setSize(d);
		GLabelImpl gLabel = new GLabelImpl();
		gLabel.setId("label");
		gLabel.setText("new1");
		gLabel.setType(DefaultTypes.LABEL);
		gLabel.setParent(gNode);
		actionDispatcher.dispatch(clientNotification.clientId, new UpdateModelAction(graph));
		

3. Open the file on which you created graph in a text editor of vscode and press cltl + s (save operation).

4. Whenever save operation is performed on the file on which we generated graph, we are triggering update model action on the diagram.

**Comment by tortmayr** [comment, marked as answer]
Hi @sathya-1994 
Short answer: To fix this issue you need to dispatch a `RequestBoundsAction` instead of an `UpdateModelAction`.
Explanation: The bounds (i.e size & width) of some components (e.g. labels) can not be computed on server-side because their actual size depends on how they are displayed & styled on client side. Therefore and update cycle in GLSP consists of the following flow:

1. Server sends a `RequestBoundsAction` to the client
2. Client computes the bounds (using sprotty's hidden rendering phase) and sends a `ComputedBoundsAction` to the server 
3. The client applies the received bound, optionally post-processes the model and then sends the `UpdateModelAction` to the client.

Steps 2 & 3 are automated, so all you have to do is send a `RequestBoundsAction` if you want to trigger a model update.

FYI: We provide the [`ModelSubmissionHandler`](https://github.com/eclipse-glsp/glsp-server/blob/c96935795bedc798dfc83c82ca46dc5a50860475/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/core/model/ModelSubmissionHandler.java) convenience class that can be used to trigger server-side updates.

**Reply by sathya-1994** [reply]
Thank you @tortmayr ,

**I used RequestBoundsAction  instead of UpdateModelAction and it fixed current issue.**
Also I am using ModelSubmissionHandler  to trigger server side updates.
